walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,100 +15,143 @@
+    meeting.
+    Sort all of events in descending order by end time.
+    Find all events that intersect with the current event and create event
+    encompassing them.
+    Continue algorithm on next event after events used.
+    Grab the start
     time for this current event and traverse through array all events that have
     an end time greater than this start time. The time slot that can not be used
     is then the end time of the current iteration and the minimum start time of
     those other events. Now take the event that after the last event that was
-    discovered still within the constraints and repeat from there. Such an
-    approach will take O(nlogn) time to sort and O(n) time to iterate. The
-    removal will take O(mn) time giving us a complexity of O(mn
+    discovered still within the constraints and repeat from there.
+
+    To handle the optional attendees case we essentially run the algorithm
+    twice, thus giving us the same overall complexity and general structure.
+
+    Complexity:
+    Such an approach will take O(nlogn) time to sort and O(n) time to iterate
+    where n is the number of meetings and m is the number of people in the
+    new meeting. The removal will take O(mn) time giving us a complexity of O(mn
     + nlogn).
     */
 
     // Store the times when we cannot meet, essentially take the inverse of this
     // to determine the times we can meet.
-
     Collection<String> meetingAttendees = request.getAttendees();
-    ArrayList<TimeRange> fullTimes = new ArrayList<>();
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
     Collection<TimeRange> toReturn = new ArrayList<>();
     // Change to ArrayList so we can sort.
     ArrayList<Event> eventList = new ArrayList<>(events);
     // Remove an event if none of the members in the meeting attended it.
     eventList.removeIf(event
                        -> !event.getAttendees().stream().anyMatch(
-                           attendee -> meetingAttendees.contains(attendee)));
+                           attendee
+                           -> meetingAttendees.contains(attendee) ||
+                                  optionalAttendees.contains(attendee)));
     Collections.sort(eventList, ORDER_END.reversed());
+    ArrayList<Event> noOptional = new ArrayList<>(eventList);
+    // Remove an event if it is only attended by optional guests.
+    noOptional.removeIf(event
+                        -> event.getAttendees().stream().allMatch(
+                            attendee -> optionalAttendees.contains(attendee)));
+    // Change this to get overlapping events.
+    ArrayList<TimeRange> timesWithOptional = getUnavailableTimes(eventList);
+    ArrayList<TimeRange> timesWithoutOptional = getUnavailableTimes(noOptional);
+    toReturn = getInverse(timesWithOptional, request);
+    if (toReturn.isEmpty() && !timesWithoutOptional.isEmpty()) {
+      toReturn = getInverse(timesWithoutOptional, request);
+    }
```

Comment:
Nice refactoring, it is much easier to read this way. I second Jie's suggestions to carefully name and comment your helper functions, which will make it even easier to read the top level function without first having read the helpers.